app/sources/HFTclTemplateController.m

```diff
@@ -65,8 +66,65 @@
     command_uint16_bits,
     command_uint32_bits,
     command_uint64_bits,
+    command_hf_min_version_required,
 };
 
+bool parseVersionString(const char* s, long* major, long* minor, long* patch) {
+    if (!s || !major || !minor || !patch) return false;
+
+    *major = *minor = *patch = 0;
+
+    const char* first = s;
+    const char* last = first + strlen(s);
+    char* end = NULL;
+
+    *major = strtol(first, &end, 10);
+    if (end == last) return false;
+    first = ++end;
+
+    *minor = strtol(first, &end, 10);
+    if (end == last) return true;
+    first = ++end;
+
+    *patch = strtol(first, &end, 10);
+
+    return true;
+}
+
+long versionInteger(long major, long minor, long patch) {
+    HFASSERT(major > 0);
+    HFASSERT(minor > 0 && minor < 999);
+    HFASSERT(patch > 0 && patch < 99);
+
+    return major * 100000 + minor * 100 + patch;
+}
+
+#define HF_XSTR(s) HF_STR(s)
+#define HF_STR(s) #s
+static const char* kVersionString = HF_XSTR(HEXFIEND_VERSION);
+#undef HF_STR
+#undef HF_XSTR
+
+long haveVersion(void) {
+    static bool inited_s = false;
```

Comment:
use Objective-C style BOOL, with NO and YES.

Reply:
Done. My forte is in C++, and if I'm in ObjC at all, it's ObjC++. In C++ the use of `static` holds the same guarantees as `dispatch_once`, so that's where all these habits are coming from.